### PR TITLE
Added ability to work with 'legacy' System.Configuration in Mono

### PR DESCRIPTION
**Pull request by steveisok**
Modified corefx System.Configuration.ConfigurationManager to be Mono machine.config aware and work with System.Configuration.

### CODE REVIEW

**Comment by marek-safar** [comment]
@steveisok what's motivation behind this change? We won't be using this code for a while

**Comment by marek-safar** [comment]
I'd be very surprised to see the whole assembly switching to CoreFX version with just these changes but let's see. I made few comments anyway

File: src/System.Configuration.ConfigurationManager/src/System/Configuration/ClientConfigurationHost.cs [hunk 1]
```diff
@@ -2,6 +2,7 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System;
```
Comment (marek-safar) [inline src/System.Configuration.ConfigurationManager/src/System/Configuration/ClientConfigurationHost.cs]: Unnecessary change

File: src/System.Configuration.ConfigurationManager/src/System/Configuration/ImplicitMachineConfigHost.cs [hunk 2]
```diff
@@ -23,14 +23,14 @@ internal ImplicitMachineConfigHost(IInternalConfigHost host)
             out string locationConfigPath, IInternalConfigRoot configRoot, params object[] hostInitConfigurationParams)
         {
             // Stash the filemap so we can see if the machine config was explicitly specified
-            _fileMap = (ConfigurationFileMap)hostInitConfigurationParams[0];
+            _fileMap = hostInitConfigurationParams[0] as ConfigurationFileMap;
```
Comment (marek-safar) [inline src/System.Configuration.ConfigurationManager/src/System/Configuration/ImplicitMachineConfigHost.cs]: Please upstream the change if it really works (the value can be null)

File: src/System.Configuration.ConfigurationManager/src/System/Configuration/Internal/WriteFileContext.cs [hunk 3]
```diff
@@ -13,14 +13,22 @@ internal class WriteFileContext
         private const int SavingRetryInterval = 100;    // 100 milliseconds
         private readonly string _templateFilename;
 
+#if !MONO
```
Comment (marek-safar) [inline src/System.Configuration.ConfigurationManager/src/System/Configuration/Internal/WriteFileContext.cs]: Should be handled at using level with #if CODEDOM

File: src/System.Configuration.ConfigurationManager/src/System/Configuration/TypeUtil.cs [hunk 4]
```diff
@@ -135,6 +141,23 @@ internal static Type GetType(IInternalConfigHost host, string typeString, bool t
             return type;
         }
 
+#if MONO
```
Comment (marek-safar) [inline src/System.Configuration.ConfigurationManager/src/System/Configuration/TypeUtil.cs]: No need to push it to CoreFX, make the type partial and keep it in Mono if necessary

File: src/System.Configuration.ConfigurationManager/src/System/Configuration/ImplicitMachineConfigHost.cs [hunk 5]
```diff
@@ -51,17 +51,29 @@ public override string GetStreamName(string configPath)
 
         public override Stream OpenStreamForRead(string streamName)
         {
+#if MONO
```
Comment (marek-safar) [inline src/System.Configuration.ConfigurationManager/src/System/Configuration/ImplicitMachineConfigHost.cs]: Could you upstream it without the ugly #if-defs?

File: src/System.Configuration.ConfigurationManager/src/System/Configuration/ImplicitMachineConfigHost.cs (hunk 2 above)
Comment (steveisok) [reply src/System.Configuration.ConfigurationManager/src/System/Configuration/ImplicitMachineConfigHost.cs]: OK - the interesting part here is that System.Web has a different order to hostInitConfigurationParams.  Hence the need to use as.

File: src/System.Configuration.ConfigurationManager/src/System/Configuration/TypeUtil.cs (hunk 4 above)
Comment (steveisok) [reply src/System.Configuration.ConfigurationManager/src/System/Configuration/TypeUtil.cs]: The #if MONO is confusing here... 

If you were to take corefx sys.config and compare it against refsrc System.Configuration, you would have the following files missing:

ConfigurationBuilder.cs
ConfigurationBuilderChain.cs
ConfigurationBuilderSettings.cs
ConfigurationBuildersSection.cs
ConfigurationStringConstants.cs
Internal/IInternalConfigurationBuilderHost.cs

ConfigurationBuildersSection.cs actually uses these methods...  Which means you could bridge corefx sys.config across the board (mono, .net 4.7, etc).  

That said, I have no problem making it partial and within mono.

File: src/System.Configuration.ConfigurationManager/src/System/Configuration/ImplicitMachineConfigHost.cs (hunk 5 above)
Comment (akoeplinger) [reply src/System.Configuration.ConfigurationManager/src/System/Configuration/ImplicitMachineConfigHost.cs]: I don't think we can upstream this, the embedded machine.config only makes sense on Mono.

Comment (steveisok) [reply src/System.Configuration.ConfigurationManager/src/System/Configuration/ImplicitMachineConfigHost.cs]: The only way it could be upstreamed is if corefx allowed an embedded machine.config.  From what I understand, that is a requirement for Xamarin.

Comment (marek-safar) [reply src/System.Configuration.ConfigurationManager/src/System/Configuration/ImplicitMachineConfigHost.cs]: I think we support both, the default and embedded modes this code picks just one. I think upstream would be ok to take the change if done little bit more cleaner

Comment (akoeplinger) [reply src/System.Configuration.ConfigurationManager/src/System/Configuration/ImplicitMachineConfigHost.cs]: > I think we support both, the default and embedded modes this code picks just one

I'm confused by this. The code in this PR supports both embedded and on-disk machine.config (`streamName` will be the full path to machine.config in this case).

Comment (marek-safar) [reply src/System.Configuration.ConfigurationManager/src/System/Configuration/ImplicitMachineConfigHost.cs]: Do we really throw when we cannot find machine config?

Comment (steveisok) [reply src/System.Configuration.ConfigurationManager/src/System/Configuration/ImplicitMachineConfigHost.cs]: Yeah, the current mono System.Config does.

**Comment by steveisok** [comment]
@marek-safar changes updated.

File: src/System.Configuration.ConfigurationManager/src/System/Configuration/Internal/WriteFileContext.cs [hunk 6]
```diff
@@ -4,6 +4,11 @@
 
 using System.IO;
```
Comment (marek-safar) [inline src/System.Configuration.ConfigurationManager/src/System/Configuration/Internal/WriteFileContext.cs]: Could you upstream this change?

Comment (steveisok) [reply src/System.Configuration.ConfigurationManager/src/System/Configuration/Internal/WriteFileContext.cs]: @marek-safar the change has been upstreamed.
